<?xml version="1.0"?>
<robot
    xmlns:xacro="http://www.ros.org/wiki/xacro" name="assemble_dual_quadrotors_male_mag" >

  <!-- general attribute -->
  <baselink name="fc" />
  <thrust_link name="thrust" />



  <!-- kinematics [m] -->

  <xacro:property name="pos_bound" value="${0.22628}" />
  <xacro:property name="center_dis" value="${0.393}" />
  <xacro:property name="rotor_z" value="${0.052}" />
  <xacro:property name="opt_pich_1" value="${0.45}" />
  <xacro:property name="opt_yaw_1" value="${0.73108}" />
  <xacro:property name="opt_pich_2" value="${0.523535}" />
  <xacro:property name="opt_yaw_2" value="${-2.08113}" />
  <xacro:property name="opt_pich_3" value="${0.523535}" />
  <xacro:property name="opt_yaw_3" value="${2.08113}" />
  <xacro:property name="opt_pich_4" value="${0.45}" />
  <xacro:property name="opt_yaw_4" value="${-0.73108}" />


  <!-- dynamics -->
  <m_f_rate value="-0.006" /> <!-- drag torque rate, the true value should be -0.006 -->
  <xacro:property name="max_force" value="15.0" /> <!-- [N] -->
  <xacro:property name="min_force" value="0.0" /> <!-- [N] -->

  <!-- friction -->
  <xacro:macro name="damping_factor" params="link">
    <gazebo reference="main_body">
      <dampingFactor>0.00</dampingFactor>
    </gazebo>
  </xacro:macro>

  <xacro:macro name="extra_module" params="name parent *origin *inertial visible:=0 model_url:=1 scale:=1">
    <joint name="${parent}2${name}" type="fixed">
      <parent link="${parent}"/>
      <child link="${name}"/>
      <xacro:insert_block name="origin" />
    </joint>
    <link name="${name}">
      <xacro:insert_block name="inertial" />
      <xacro:if value="${visible == 1}">
        <visual>
          <origin rpy="0 0 0" xyz="0 0 0"/>
          <geometry>
            <mesh filename="${model_url}" scale= "${scale} ${scale} ${scale}"/>
          </geometry>
        </visual>
      </xacro:if>
    </link>
  </xacro:macro>

  <xacro:macro name="propeller_module" params="id x y z pich yaw direction">
    <joint name="rotor${id}" type="continuous">
      <limit effort="100.0" lower="${min_force}" upper="${max_force}" velocity="0.5"/>
      <parent link="main_body"/>
      <child link="thrust${id}"/>
      <origin rpy="0 ${pich} ${yaw}" xyz="${x} ${y} ${z}"/>
      <axis xyz="0 0 ${direction}"/>
    </joint>

    <link name="thrust${id}">
      <!-- visual & collisiont -->
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="0.0052"/>
        <inertia
            ixx="0.00000807" ixy="0.0" ixz="0.0"
            iyy="0.00000807" iyz="0.0"
            izz="0.00001612"/>
      </inertial>
      <visual>
        <origin rpy="0 0 0" xyz="0 0 0"/>
        <geometry>
          <xacro:if value="${direction == 1}">
            <mesh filename="package://assemble_quadrotors/urdf/mesh/5_inchi_propeller.STL" />
          </xacro:if>
          <xacro:unless value="${direction == 1}">
	    <mesh filename="package://assemble_quadrotors/urdf/mesh/5_inchi_propeller.STL" />
          </xacro:unless>
        </geometry>
      </visual>
    </link>
    <xacro:damping_factor link ="thrust${id}"/>

    <transmission name="rotor_tran${id}">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="rotor${id}">
        <!-- TODO: effort is torque, maybe position is enough -->
        <hardwareInterface>RotorInterface</hardwareInterface>
      </joint>
      <actuator name="rotor${id}">
        <!-- TODO: effort is torque, maybe position is enough -->
        <hardwareInterface>RotorInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>

  </xacro:macro>

  <!-- dummy root link for KDL -->
  <link name="root">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <mass value="0.00001"/>
      <inertia
          ixx="0.000001" ixy="0.0" ixz="0.0"
          iyy="0.000001" iyz="0.0"
          izz="0.000002"/>
    </inertial>
  </link>
  <joint name="root_joint" type="fixed">
    <parent link="root"/>
    <child link="main_body"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <axis xyz="0 0 1"/>
  </joint>

  <!-- main_body -->
  <link name="main_body">
    <inertial>
      <mass value = "1.23" />
      <origin xyz="0.00218 -0.00165 -0.00150" rpy="0 0 0"/>
    <inertia
        ixx="0.00626" ixy="0.00002959" ixz="-0.000716"
        iyy="0.0152" iyz="-0.00004101"
        izz="0.0187"/>
    </inertial>

    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://assemble_quadrotors/urdf/mesh/Assemble_quadrotor_male_magnet.STL"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.3525 0.26 0.092" />
      </geometry>
    </collision>
  </link>

  <!-- spinal -->
  <xacro:extra_module name = "fc" parent = "main_body" visible = "0" >
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <inertial>
      <mass value = "0.05" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia
          ixx="0.00001" ixy="0.0" ixz="0.0"
          iyy="0.00001" iyz="0.0"
          izz="0.00002"/>
    </inertial>
  </xacro:extra_module>

  <!-- propellers -->
  <xacro:propeller_module id = "1" x = "-0.10754" y = "-0.12284" z = "0.05287" pich = "${-opt_pich_3}" yaw = "${opt_yaw_3}" direction = "1" />
  <xacro:propeller_module id = "2" x = "0.10610" y = "-0.10682" z = "0.05662" pich = "${-opt_pich_4}" yaw = "${opt_yaw_4}"  direction = "-1" />
  <xacro:propeller_module id = "3" x = "0.10618" y = "0.10708" z = "0.05674"  pich = "${-opt_pich_1}" yaw = "${opt_yaw_1}" direction = "1" />
  <xacro:propeller_module id = "4" x = "-0.10744" y = "0.12274" z = "0.05289"  pich = "${-opt_pich_2}" yaw = "${opt_yaw_2}"  direction = "-1" />

  
  ##### dummy grasp links ####
  <link name="dummy_arm">
    <inertial>
      <mass value = "0.01" />
      <origin xyz="0.0 0 0" rpy="0 0 0"/>
      <inertia
          ixx="0.0001" ixy="0.0000" ixz="0.0000"
          iyy="0.0001" iyz="0.0000" izz="0.0001"/>
    </inertial>
    <visual>
      <geometry>
        <box size="0.04 0.04 0.04"/>
      </geometry>
    </visual>
  </link>
  <joint name="base_dummy_arm_joint" type="fixed">
    <parent link="main_body"/>
    <child link="dummy_arm"/>
    <origin xyz="0.2 0.0 -0.04" rpy="0 0 0"/>
  </joint>
  <joint name="dummy_arm_palm_joint" type="revolute">
    <parent link="dummy_arm"/>
    <origin xyz="0.02 0 0" rpy="0 0 0"/>
    <child link="dummy_palm"/>
    <axis xyz="1 0 0"/>
    <limit lower="0" upper="0" effort="0" velocity="0"/>
    <dynamics damping="0.2" friction="0.02"/>
  </joint>

  <link name="dummy_palm">
    <inertial>
      <mass value = "0.001" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia
          ixx="0.0001" ixy="0.0000" ixz="0.0000"
          iyy="0.0001" iyz="0.0000" izz="0.0001"/>
    </inertial>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.04 0.04 0.04"/>
      </geometry>
    </collision>
  </link>

  <joint name="dummy_palm_gripper_joint" type="revolute">
    <parent link="dummy_palm"/>
    <child link="dummy_gripper"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <axis xyz="1 0 0"/>
    <limit lower="0" upper="0" effort="0" velocity="0"/>
    <dynamics damping="0.2" friction="0.02"/>
  </joint>


  <link name="dummy_gripper">
    <inertial>
      <mass value = "0.001" />
      <origin xyz="0.0 0 0" rpy="0 0 0"/>
      <inertia
          ixx="0.0001" ixy="0.0000" ixz="0.0000"
          iyy="0.0001" iyz="0.0000" izz="0.0001"/>
    </inertial>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.04 0.04 0.04"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 0 0" rpy="0.0 0 0"/>
      <geometry>
        <box size="0.04 0.04 0.04"/>
      </geometry>
    </visual>
  </link>
  
  <transmission name="joint_tran1">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="dummy_arm_palm_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="servo1">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="joint_tran2">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="dummy_palm_gripper_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="servo2">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

</robot>
